packages/aws-cdk-lib/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -424,7 +425,7 @@ export abstract class InitFile extends InitElement {
   public static fromAsset(targetFileName: string, path: string, options: InitFileAssetOptions = {}): InitFile {
     return new class extends InitFile {
       protected _doBind(bindOptions: InitBindOptions) {
-        const asset = new s3_assets.Asset(bindOptions.scope, `${targetFileName}Asset`, {
+        const asset = new s3_assets.Asset(bindOptions.scope, `${md5hash(bindOptions.scope.node.children.toString())}${targetFileName}Asset`, {
```

Comment:
Personally I'd find a comment useful here explaining why the hash is being taken this way.